### ISSUE TITLE
Proposal - move to JNA

### ISSUE THREAD

**Issue by zafarella** [techblue/jmagick#40]
As an alternative to existing calls, following could be used to make call to native libs https://github.com/java-native-access/jna

This will address many versioning limitation and should avoid many memory leek issues, but will keep the "DSL" as is.

Open to discussion/suggestions. Just a thought. Feel free to add your opinion/experience

**Comment by swrdfish** [comment]
@nordfalk I would like to volunteer. I was trying to wrap imagemagick with JNA for a project, and I am thinking of contributing. But I need some help as I am pretty new to JNA and I am stuck at some basic stuff (I am getting segmentation fault while trying to retrieve the Image struct). I have hosted the code here: https://github.com/swrdfish/dumbledore

Can anyone point me to some documentation or give some suggestions to resolve such problems?

**Comment by nordfalk** [comment]
Looks like you have made a lot of progress, @swrdfish .

**Comment by nordfalk** [comment]
Congratulations, @swrdfish !
Looking at https://github.com/swrdfish/dumbledore/blob/master/src/main/java/com/swrdfish/tools/dumbledore/Demo.java my impression is that those that are currently using jMagick will need to rewrite their client code - would it be possible possible to minimize how much the client code would need to be rewritten?
Perhaps by using some Java wrapper classes that invoke your JNA layer.

**Comment by swrdfish** [comment]
@nordfalk yes using jMagick is much simpler that using direct JNA bindings. So having a wrapper will definitely help. Also the current version has a lot of small memory issues. So we need tests to cover atleast the core functionality.

I am not sure what the API should look like, maybe I can try to clone the jMagick interface, that will make porting current client code easy. Any suggestions?

**Comment by nordfalk** [comment]
Yes, try to clone the jMagick interface, that will make porting current
client code easy.

Sorry for the lateral reply

tor. d. 9. aug. 2018 07.03 skrev Binayak Ghosh <notifications@github.com>:

**Comment by judos** [comment]
@swrdfish i'm interested in your jna interface, did you get it working? As far as I can see only the Demo code exists, but you didn't adapt it for the whole api here in jmagick right? Also need it for a project.

**Comment by swrdfish** [comment]
@judos I never got it working properly, and no I did not adapt it for the whole api.

**Comment by hantsy** [comment]
The native access will protected in future, https://openjdk.org/jeps/472
